### ISSUE TITLE
Basic Block in function does not have terminator when using custom alloc crate.

### ISSUE THREAD

**Issue by tbelaire** [rust-lang/rust#33065]
I am getting a LLVM error after implementing an `allocator` crate and using `Box`.

```
rustc 1.10.0-dev (576229fea 2016-04-15)

---
$ cargo build --target=x86_64-unknown-none-gnu
...
Basic Block in function 'rust_main' does not have terminator!
label %resume
LLVM ERROR: Broken function found, compilation aborted!
error: Could not compile `hello`.

```

I've stripped down the example to a quite small 3 crate example:

https://github.com/tbelaire/rust.ko/tree/rot13-minimal-example

```
// src/lib.rs
let x = alloc::boxed::Box::new(32);      
```

seems to be the offending line, though it needs to have another statement after it for it to fail to compile.

`std/kmalloc_allocator/src/lib.rs` is a do nothing `#![allocator]` crate, and then `liballoc` from the stdlib.

Oh, and I have a target file, otherwise this also compiles.

The rustc I am using was compiled with `./configure --enable-debug-assertions`, to get more information from the LLVM crash.

**Comment by alexcrichton** [comment]
Thanks for the report! Could you summarize the steps to take to reproduce this as well?

**Comment by tbelaire** [comment]
```
git clone git@github.com:tbelaire/rust.ko
cd rust.ko
git checkout rot13-minimal-example
cargo build --target=x86_64-unknown-none-gnu
```
Should be sufficient to reproduce it, though you will get an error message like:
```
Assertion failed: ((Args.size() == FTy->getNumParams() || (FTy->isVarArg() && Args.size() > FTy->getNumParams())) && "Calling a function with bad signature!"), function init, file /Users/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-mac/build/src/llvm/lib/IR/Instructions.cpp, line 240.
```
if you are using a normal build of Rust instead of one with the extra assertions turned on.

**Comment by tbelaire** [comment]
Ahg, this might be hard to reproduce.  I just tried it on a different linux box and the minimal example wasn't failing anymore, but the main rot13 branch still was.  This is with nightly 04-10, 04-20, 04-25 all showing that behaviour on my digital ocean droplet.

Does the minimal example compile on other people's machines?

**Comment by johalun** [comment]
I might have the same or similar problem described here: 
https://users.rust-lang.org/t/no-std-write-fmt-macros-error/9255

I'm writing a FreeBSD kernel module in Rust (inspired by https://github.com/tsgates/rust.ko). I use Xargo to build a custom core. I have a minimal implementation of std, like rust.ko. 

This error happens whenever I try to use write_fmt, using lazy_static or trying to create an static object like spin::Mutex (primitives are ok).

**Comment by johalun** [comment]
Test case: https://github.com/johalun/llvm-test-bug

**Comment by Mark-Simulacrum** [comment]
The llvm-test-bug repo appears to work for me today, and the other example here doesn't compile due to, I think, using an old libcore clone/copy that doesn't compile today. Closing.

**Comment by johalun** [comment]
Thanks for confirming this!
